pkg/client/client_venafi_cloud.go

```diff
@@ -152,6 +154,57 @@ func (c *VenafiSvcAccountCredentials) IsClientSet() bool {
 	return c.ClientID != "" && c.PrivateKeyFile != ""
 }
 
+// PostDataReadingsWithOptions uploads the slice of api.DataReading to the Venafi Cloud backend to be processed.
+// The Options are then passed as URL params in the request
+func (c *VenafiCloudClient) PostDataReadingsWithOptions(readings []*api.DataReading, opts Options) error {
+	payload := api.DataReadingsPost{
+		AgentMetadata:  c.agentMetadata,
+		DataGatherTime: time.Now().UTC(),
+		DataReadings:   readings,
+	}
+	data, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+
+	if !strings.HasSuffix(c.uploadPath, "/") {
+		c.uploadPath = fmt.Sprintf("%s/", c.uploadPath)
+	}
+
+	venafiCloudUploadURL, err := url.Parse(filepath.Join(c.uploadPath, c.uploaderID))
+	if err != nil {
+		return err
+	}
+
+	// validate options and send them as URL params
+	query := venafiCloudUploadURL.Query()
+	stripHTML := bluemonday.StrictPolicy()
+	if opts.ClusterName != "" {
+		query.Add("name", stripHTML.Sanitize(opts.ClusterName))
+	}
```

Comment:
(Passing by) Just for future reference, I was wondering this: why do we pass the query param `name` set to the config's `cluster_id`? The `cluster_id` is already present in the request body of `/v1/tlspk/upload/clusterdata/:uploaderId`, more specifically under `agent_metadata`. For example:

```http
POST https://api.venafi.cloud/v1/tlspk/upload/clusterdata/no HTTP/2.0

content-type: application/json
authorization: Bearer ...
content-length: 536
accept-encoding: gzip
user-agent: Go-http-client/2.0
{
  "agent_metadata": {
    "version": "development",
    "cluster_id": "mvalais Kind Cluster"
  },
  "data_gather_time": "2024-07-26T13:03:39.831129Z",
  "data_readings": [
    {
      "cluster_id": "mvalais Kind Cluster",
      "data-gatherer": "k8s-discovery",
      "timestamp": "2024-07-26T15:03:09+02:00",
      "data": {
        "server_version": {
          "major": "1",
          "minor": "29",
          "gitVersion": "v1.29.2",
          "gitCommit": "4b8e819355d791d96b7e9d9efe4cbafae2311c88",
          "gitTreeState": "clean",
          "buildDate": "2024-02-14T22:25:42Z",
          "goVersion": "go1.21.7",
          "compiler": "gc",
          "platform": "linux/arm64"
        }
      },
      "schema_version": "v2.0.0"
    }
  ]
}
```

When I don't pass the query param `name=mvalais+Kind+Cluster`, I get the following error:

```json
[{"message":"could not find a Kubernetes cluster linked to the client in the Venafi Control Plane"}
```

Reply:
The payload body could be quite large, several MB depending on the cluster where the agent is installed, in the backend entrypoint we wanted to be able to get this information without having to parse the entire upload before the actual processing pipeline kicks in. (Note: we could stream the upload JSON but we thought query params would work well instead).

The error can be improved, but essentially you have 2 options to onboard a cluster:
* venctl installed which creates a service account and cluster_connection (relation of cluster and service account) in VCP.
* Helm installation which requires the backend entrypoint to create a cluster_connection in VCP for your cluster and service account.

The error is saying that the cluster and service account are not linked in VCP.

In the venafi docs we tell users following the helm installation that they are **required** to set the cluster name, which will also be present in the UI. We could introduce an `auto-onboard` field to the config but since the cluster name field is required for the helm installation we thought we could use it rather than introducing another field.

If the cluster name is set in the config it will be used in the query param. The user wouldn't interact with the query params.

Reply:
Thanks, that a super useful insight!